lib/common/JSONImporter/NodeStatePatch.ts

```diff
@@ -0,0 +1,519 @@
+import {NodeSelections} from './NodeSelectors';
+import {NodeChangeSet} from './NodeChangeSet';
+import {NodeSearchUtils, setNested} from './Utils';
+import {Maybe, Result} from 'ts-monads';
+import diff, {ChangeSet, ChangeType} from 'changeset';
+import JSONImporter from "../JSONImporter";
+import Core = GmeClasses.Core;
+
+type PatchFunction = (node: Core.Node, change: NodeChangeSet, resolvedSelectors: NodeSelections) => Promise<Result<PatchResult, PatchError>>;
+type PatchResultPromise = Promise<Result<PatchResult, PatchError>>;
+
+class PatchResult {
+    patches: NodeChangeSet | NodeChangeSet[];
+
+    constructor(patches: NodeChangeSet | NodeChangeSet[]) {
+        this.patches = patches;
+    }
+}
+
+export class PatchError extends Error {
+    constructor(msg: string) {
+        super(msg);
+    }
+}
+
+export interface PatchOperation {
+    core: GmeClasses.Core;
+    put: PatchFunction;
+    delete: PatchFunction;
+}
+
+export abstract class NodeStatePatch implements PatchOperation {
+    core: GmeClasses.Core;
+    nodeSearchUtils: NodeSearchUtils
+
+    constructor(core: GmeClasses.Core, nodeSearchUtils: NodeSearchUtils) {
+        this.core = core;
+        this.nodeSearchUtils = nodeSearchUtils;
+    }
+
+    keyLengthValidator(change) {
```

Comment:
Neither would this (at least not on the base class where I am not convinced it makes the most sense).

Reply:
Kept it for reuse in couple of places